Bluetooth: controller: Add option for ticker to return expiration info

Tronil opened the pull request:
It is now possible (via the extended ticker interface) to ask for expiration information for another ticker ID to be returned in the ticker callback. This utilises the extended ticker interface and a different callback function to keep the current ticker interface as unchanged as possible. Since extended advertising makes use of this feature, the extended ticker interface is now selected when extended advertising is supported.

Updated the 3 places in extended advertising where the ticker was previously inquired about the expiration of another ticker ID in a separate mayfly (adv_aux, adv_sync and adv_iso).

Signed-off-by: Troels Nilsson <trnn@demant.com>

Tronil commented:
@cvinayak @mtpr-ot the proposed changes that we discussed to avoid having to fire off a mayfly to get the offsets for auxptr and friends that may or may not run in time.

Feedback is very welcome

cvinayak reviewed (changes requested):
Not finished reviewing fully. But you can address some of the comments.

mtpr-ot reviewed (commented):
I would prefer to have the ticker change and the adv changes int separate commits.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1067,24 +1112,57 @@ void ticker_worker(void *param)
 			DEBUG_TICKER_TASK(1);
 			/* Invoke the timeout callback */
-			ticker->timeout_func(ticks_at_expire,
+
+#if defined(CONFIG_BT_TICKER_EXT)
+			if (ticker->ext_data && ticker->ext_data->ext_timeout_func) {
+				struct ticker_expire_info other_expire_info;
+				struct ticker_ext_timeout_data data;
+				bool found = false;
+
+				if (ticker->ext_data->expire_info_id != TICKER_NULL) {
+					found = ticker_get_expire_info(instance,
+							ticker->ext_data->expire_info_id,
+							ticker->next, &other_expire_info);
+				}
+
+				data.ticks_at_expire = ticks_at_expire;
+				data.ticks_drift = ticks_drift;
+				data.remainder = ticker->remainder_current;
+				data.lazy = must_expire_skip ?
+					    TICKER_LAZY_MUST_EXPIRE :
+					    ticker->lazy_current;
+				data.force = ticker->force;
+				data.other_expire_info = found ? &other_expire_info : NULL;
+
+				ticker->ext_data->ext_timeout_func(&data, ticker->context);
+			} else
+#endif /* CONFIG_BT_TICKER_EXT */
```
I assume you mean defer to `ticker_job`, and obtain the information for the next timeout, then present/provide the information in `ticker_worker`? This could work if it's acceptable to get information which is not 100% up to date, but how it looked at the last `ticker_job` run. There would have to be code running for every invocation of `ticker_job`, which manipulates the nodes.

Tronil replied:
> Due to the O(n) nature of `ticker_get_expire_info`, we should perform this query in the `ticker_job` when the head ticker is a ticker_ext with `expire_info_id`. The `ticker_expire_info` be a member in `ext_data`.

That would work as well, but as @mtpr-ot mentions it would mean that we will need to fire off re-calculations whenever something in the ticker changes (which is potentially expensive - I don't have a good enough overview to know how expensive in practice though). It would add more complexity for certain though.

The question is if the added overhead of this calculation is bad enough to warrant that; There is an upper bound limited by the number of valid ticker IDs, so we can be sure it won't grow beyond a certain point. In our case, the worst-case of having to search through all valid ticker IDs will give an overhead of ~26 microseconds - which should be reasonable. Of course, if you have a ticker with something like 255 valid ticker IDs it could be problematic (though you could adjust the EVENT_OVERHEAD_START_US value to compensate) - but is that realistic?

cvinayak replied:
> I assume you mean defer to ticker_job, and obtain the information for the next timeout, then present/provide the information in ticker_worker? This could work if it's acceptable to get information which is not 100% up to date, but how it looked at the last ticker_job run. There would have to be code running for every invocation of ticker_job, which manipulates the nodes.

The calculation will be done before the ticker compare is setup in the `ticker_job`, that is, only when the head in the ticker list changes to being the ticker that needs the expire info. In timeline, execution of `ticker_get_expire_info()` followed by call to `ext_timeout_func()` or `ticker_timeout_func ()` follow the same order with the advantage that `ticker_worker` consumes lower CPU time at its execution priority level.

> That would work as well, but as @mtpr-ot mentions it would mean that we will need to fire off re-calculations whenever something in the ticker changes (which is potentially expensive - I don't have a good enough overview to know how expensive in practice though). It would add more complexity for certain though.

In practice, head does not change until the ticker expire unless `ticker_start`, `ticker_update`, or `ticker_stop` (that change the head) is called between the compare being set and ticker expires.

Tronil replied:
> The calculation will be done before the ticker compare is setup in the `ticker_job`, that is, only when the head in the ticker list changes to being the ticker that needs the expire info. In timeline, execution of `ticker_get_expire_info()` followed by call to `ext_timeout_func()` or `ticker_timeout_func ()` follow the same order with the advantage that `ticker_worker` consumes lower CPU time at its execution priority level.

The problem with only running the calculation for the head is that you can have several tickers expire within the same ticker_worker call. In most cases it will be fine to just have the information prepared in advance for the head; But in cases where two (or more) tickers expire in the same call.

We have a couple of ways we can deal with this:

1. Try to determine in advance which tickers could be expiring in the next call to ticker_worker instead of only looking at the head. I think we would need some kind of fudge factor here, since it will vary across hardware and load how close tickers will have to be to risk expiring in the same call.
2. Do the calculation for the head only, and in the unlikely case we end up expiring more than the head *and* it/they need expiration info, we run the calculation inside ticker_worker. This would still mean that we would sometimes have some delay before the ticker_cb is called though, so maybe it doesn't make sense
3. Always calculate expiration info for all tickers that requires it, not just the head. This is potentially very expensive though, since you can potentially have a lot of them

Tronil commented on subsys/bluetooth/controller/ticker/ticker.h:
```diff
@@ -101,6 +109,26 @@ typedef void (*ticker_timeout_func) (uint32_t ticks_at_expire,
 				     uint16_t lazy, uint8_t force,
 				     void *context);
 
+
+struct ticker_expire_info {
+	uint32_t ticks_to_expire;
+	uint32_t remainder;
+	uint16_t lazy;
+};
+
+struct ticker_ext_timeout_data {
+	uint32_t ticks_at_expire;
+	uint32_t ticks_drift;
+	uint32_t remainder;
+	uint16_t lazy;
+	uint8_t force;
+	struct ticker_expire_info *other_expire_info;
+};
+
+/** \brief Timer extended timeout function type.
+ */
+typedef void (*ticker_ext_timeout_func) (const struct ticker_ext_timeout_data *data, void *context);
+
```
That would work as well. I assume you mean that we can then use the same ticker_timeout_func callback function as before instead of having a different one; Extending the context is probably a bit cleaner code-wise

cvinayak replied:
yes. As we know the `ticker_timeout_func` was supplied using the extended interface hence the `context` is always `ticker_ext_context`.

Tronil commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -825,7 +825,6 @@ config BT_TICKER_NEXT_SLOT_GET_MATCH
 
 config BT_TICKER_EXT
 	bool "Ticker extensions"
-	depends on !BT_TICKER_LOW_LAT && !BT_TICKER_SLOT_AGNOSTIC
```
Does `BT_TICKER_LOW_LAT` imply that extended advertising is not supported? Because it will now select `BT_TICKER_EXT`

cvinayak replied:
`BT_TICKER_EXT` introduces the slot_window feature (nudging the ticker to upto slot_window when the ticker overlaps with a non-slot window ticker) which is only implemented for `!BT_TICKER_LOW_LAT`.

Tronil replied:
> `BT_TICKER_EXT` introduces the slot_window feature (nudging the ticker to upto slot_window when the ticker overlaps with a non-slot window ticker) which is only implemented for `!BT_TICKER_LOW_LAT`.

I believe I changed enough of the flagging inside the ticker so you can now enable `BT_TICKER_EXT` without that feature being active (at least that was the intent). The idea is that  `BT_TICKER_EXT` means that the *_ext functions are available, while you can still use `BT_TICKER_LOW_LAT` to disable slot windows.

Tronil commented on subsys/bluetooth/controller/ll_sw/lll_adv.h:
```diff
@@ -127,12 +131,26 @@ struct lll_adv_aux {
 	 */
 	uint16_t data_chan_counter;
 
+	uint16_t data_chan_id;
+	struct {
+		uint8_t data_chan_map[PDU_CHANNEL_MAP_SIZE];
+		uint8_t data_chan_count:6;
+	} chm[DOUBLE_BUFFER_SIZE];
+	uint8_t  chm_first;
+	uint8_t  chm_last;
+
```
Two reasons:
1 - to align with lll_adv_sync/lll_adv_iso which has the channel map in the LLL structure
2 - since the channel map update and channel index calculation is now done in prepare, it made sense to have it in the LLL structure

As you mentioned in the meeting, the channel map update should be moved though

Tronil replied:
Moved this back to ull and made the discussed changes to channel map updates

Tronil commented:
> I would prefer to have the ticker change and the adv changes int separate commits.

Yes I will split into separate commits :+1:

Tronil committed:
Bluetooth: controller: Add option for ticker to return expiration info

It is now possible (via the extended ticker interface) to ask for
expiration information for another ticker ID to be returned in the
ticker callback. This utilises the extended ticker interface and a
callback function with a different context to keep the current ticker
interface as unchanged as possible.

Signed-off-by: Troels Nilsson <trnn@demant.com>

Tronil committed:
Bluetooth: controller: Use new ticker feature to get offsets

Updated the 3 places in extended advertising where the ticker was
previously inquired about the expiration of another ticker ID in a
separate mayfly (adv_aux, adv_sync and adv_iso). It will now use the
new ticker feature where another ticker IDs expiration info will be
added to the callback.

Channel map updates are now handled in ticker_cb and applied immediately
if the ticker is not running

Signed-off-by: Troels Nilsson <trnn@demant.com>

Tronil commented:
Updated PR to use ticker_job to run the calculations in advance.

It currently calculates all the requested offsets - an option would be to limit it to only calculate for those tickers that may expire in the next "round" of expiring tickers; But determining what limit to use in that logic is non-trivial, so I implemented the safe option for now.

Tronil commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_aux.c:
```diff
@@ -307,6 +271,12 @@ static int prepare_cb(struct lll_prepare_param *p)
 
+#if defined(CONFIG_BT_CTLR_ADV_PERIODIC)
+		if (pdu->adv_ext_ind.ext_hdr_len && pdu->adv_ext_ind.ext_hdr.sync_info) {
+			ull_adv_sync_lll_syncinfo_fill(pdu, lll);
```
The channel map update is now applied directly if the ticker is not running. If the ticker is running, the update is done in ticker_cb

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -1393,6 +1395,11 @@ uint8_t ull_adv_aux_chm_update(void)
 		aux->chm[chm_last].data_chan_count =
 			ull_chan_map_get(aux->chm[chm_last].data_chan_map);
 		aux->chm_last = chm_last;
+
+		if (!aux->is_started) {
+			/* Ticker not started yet, apply new channel map now */
+			aux->chm_first = aux->chm_last;
+		}
```
Do not modify `chm_first` in ULL execution context, instead modify only `chm_last`

Tronil replied:
As far as I remember, this was done to handle the case where we get two channel updates without the advertising running. Currently, that will fail due to this check earlier in the function:
```
		if (aux->chm_last != aux->chm_first) {
			/* TODO: Handle previous Channel Map Update being in
			 * progress
			 */
			continue;
		}
```
Shouldn't it be safe to update `chm_first` when the ticker is not running?

cvinayak replied:
Could you add a comment in the code explaining the same that it is safe to update the indices as the advertising is not currently running.

Tronil replied:
I have added a comment about that

cvinayak reviewed (dismissed):
I have not yet reviewed ticker.c yet, but have check that the change in this PR on target appear to have no regression when using as an ISO Broadcast Controller (i.e. audio usecase on a nRF5340 audio kit).

Hope to do more reviews by next week and get this merged soon.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_sync.c:
```diff
@@ -1134,17 +1160,38 @@ uint32_t ull_adv_sync_start(struct ll_adv_set *adv,
+
+#if defined(CONFIG_BT_CTLR_ADV_ISO)
+	ll_adv_sync_ticker_ext[sync_handle].ext_timeout_func = ticker_cb;
+	if (lll_sync->iso) {
+		ll_adv_sync_ticker_ext[sync_handle].expire_info_id = TICKER_ID_ADV_ISO_BASE +
+								    lll_sync->iso->handle;
```
```suggestion
	if (sync->lll.iso) {
		ll_adv_sync_ticker_ext[sync_handle].expire_info_id =
			TICKER_ID_ADV_ISO_BASE + sync->lll.iso->handle;
```

cvinayak commented:
I see a regression in the iso_broadcast and iso_receive samples in a draft PR branch I am currently working:
Draft PR **without** changes in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/55855
Draft PR **with** changes in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/55856

These are the commands to build the samples:
![image](https://user-images.githubusercontent.com/6350656/225604123-d7389b82-ddad-450c-90f9-7db085bf6171.png)
and
![image](https://user-images.githubusercontent.com/6350656/225604227-8aa35ed7-fd30-46af-b3e9-df5e9936056c.png)

These are the commands to run in babblesim:
![image](https://user-images.githubusercontent.com/6350656/225603873-2ccd0e39-7d36-4c01-83e2-c75dbdb0dccb.png)


The output without the changes in this PR:
![image](https://user-images.githubusercontent.com/6350656/225605550-b3c9f15a-d79d-4ec1-b68f-d9352536f73b.png)

The output with the changes in this PR rebased:
![image](https://user-images.githubusercontent.com/6350656/225604947-65bef89e-dec3-48f2-82a0-472c3d91b527.png)

Tronil commented:
> I see a regression in the iso_broadcast and iso_receive samples in a draft PR branch I am currently working: Draft PR **without** changes in this PR: #55855 Draft PR **with** changes in this PR: #55856

Looks like payload count is off by one with my changes; I'll have to dig around a bit  to find why that is.

cvinayak commented:
> > I see a regression in the iso_broadcast and iso_receive samples in a draft PR branch I am currently working: Draft PR **without** changes in this PR: #55855 Draft PR **with** changes in this PR: #55856
> 
> Looks like payload count is off by one with my changes; I'll have to dig around a bit to find why that is.

Cool, I would not figure out what it was as I was not having sync timeout. Do let me know if you need any assistance, I can spend some time next week to get this PR merged.

Tronil commented:
> > > I see a regression in the iso_broadcast and iso_receive samples in a draft PR branch I am currently working: Draft PR **without** changes in this PR: #55855 Draft PR **with** changes in this PR: #55856
> > 
> > 
> > Looks like payload count is off by one with my changes; I'll have to dig around a bit to find why that is.
> 
> Cool, I would not figure out what it was as I was not having sync timeout. Do let me know if you need any assistance, I can spend some time next week to get this PR merged.

Found the issue. There is a timing difference now which causes lll_adv_iso.c's `prepare` function to run just before `ull_adv_iso_lll_biginfo_fill` is called; This causes `lll->latency_prepare` to be set to a non-zero value and this screws up the payload count. As far as I can see this calculation of the payload count should not use `lll->latency_prepare`, since it is basically just a snapshot of lazy (+1) which is included already: 
```
	payload_count = lll_iso->payload_count + ((lll_iso->latency_prepare +
						   lll_iso->iso_lazy) * lll_iso->bn);
```
I'll update my PR with a fix for this - might be after the weekend though.

Tronil commented on subsys/bluetooth/controller/ll_sw/ull_adv_iso.c:
```diff
@@ -1064,71 +1053,18 @@ static void adv_iso_chm_complete_commit(struct lll_adv_iso *lll_iso)
 
-	payload_count = lll_iso->payload_count + ((lll_iso->latency_prepare +
-						   lazy) * lll_iso->bn);
+	payload_count = lll_iso->payload_count + (lll_iso->iso_lazy * lll_iso->bn);
```
Yes, it is set in the ticker_cb based on the other_expire_info that the ticker provides (in ull_adv_sync.c)

cvinayak replied:
`lll_iso->latency_prepare`  and `lll_iso->iso_lazy` should both be used, see implementation of `ull_conn_event_counter()` how it handles prepare callback that are in prepare pipeline, i.e. latency accumulated with reference get incremented.

Tronil replied:
> `lll_iso->latency_prepare` and `lll_iso->iso_lazy` should both be used, see implementation of `ull_conn_event_counter()` how it handles prepare callback that are in prepare pipeline, i.e. latency accumulated with reference get incremented.

I don't know how latency_prepare in `lll_conn` is used; But latency_prepare in `lll_adv_iso` is set to `lazy + 1` in `prepare` and then set back to 0 `prepare_cb_common` (see lll_adv_iso.c) - ie. it is just a snapshot of the lazy parameter. So if you include latency_prepare in the calculation of payload_count, you will effectively be calculating it as `lll_iso->payload_count + ((lll_iso->iso_lazy*2 + 1) * lll_iso->bn);` (when called in between `prepare` and `prepare_common`), which looks very wrong to me.

cvinayak replied:
`latency_prepare` in all state/role LLL contexts accumulate `lazy` after `ticker_cb` has executed and the prepare callbacks wait (preempt timeout duration) in the pipeline until `prepare_cb` is executed. During this wait, prepare can be aborted (removed from pipeline) and hence `latency_prepare` accumulates all `lazy` value until subsequent `prepare_cb` gets executed.

That said, I will test this PR and get back this week.

Tronil replied:
> `latency_prepare` in all state/role LLL contexts accumulate `lazy` after `ticker_cb` has executed and the prepare callbacks wait (preempt timeout duration) in the pipeline until `prepare_cb` is executed. During this wait, prepare can be aborted (removed from pipeline) and hence `latency_prepare` accumulates all `lazy` value until subsequent `prepare_cb` gets executed.

Ah - yes I missed the `+=` in `prepare`. But since it is incremented by `lazy + 1` it is problematic to include this in the payload_count calculation... I'll have a think about how this can be handled.

cvinayak replied:
> since it is incremented by lazy + 1 it is problematic to include this in the payload_count calculation

Yes, `prepare_cb` execution depends on requirement to preempt active/current radio event or immediately execute. Hence, I assume the special handling in `ull_conn_event_counter()` used by LLCP implementation.

Tronil replied:
Implemented something similar to ´ull_conn_event_counter()´ to account for being in a post-prepare situation; Looks like it is working.

cvinayak replied:
Tested Broadcast ISO samples (https://github.com/zephyrproject-rtos/zephyr/pull/55856)  and Central Multiple CIS support BabbleSIM (https://github.com/zephyrproject-rtos/zephyr/pull/54657/commits/32a128d55421ebe89fe18ffba158acd3d53edb28). The changes in this PR show no regression in these minimal testing or in the CI tests.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -21,6 +21,24 @@
+#if defined(CONFIG_BT_CTLR_ADV_ISO)
+#define TICKER_EXPIRE_INFO_MAX (BT_CTLR_ADV_AUX_SET + BT_CTLR_ADV_SYNC_SET*2)
+#else
+#define TICKER_EXPIRE_INFO_MAX (BT_CTLR_ADV_AUX_SET + BT_CTLR_ADV_SYNC_SET)
```
This means that if a new use case comes up, e.g. for CIS Central connection spacing, this would need to be changed. I think we need to add some kind of Kconfig for additional expire info entries, or skip the calculation and have a value set in Kconfig, with a value described in a comment.
We can do this later - not blocking for approval.

cvinayak replied:
I will prefer allocations for the extended features in ticker to come from caller modules. But for now, I did not have objections the way it  is now. Yes, definitely we should try move use of Controller Kconfigs inside ticker and may be have ticker specific Kconfig that can have the values inherited.

mtpr-ot reviewed (approved):
LGTM. Minor comment for later
